### PR TITLE
Support path recursion via process.mainModule.paths

### PR DESCRIPTION
**Pull request by perry-mitchell**
An attempt to address charlike/get-installed-path#11 by adding a new field to `opts`: `opts.paths`. By allowing a user to pass in an array of paths to search, `get-installed-path` can be extended to support many other methods of locating packages. I opted for the `process.mainModule.paths` example, which returns an array of node_modules paths that are _relevant_ to the current entry script.

It might contain an array of paths that looks like the following:

```javascript
[
    "/home/user/git/main/node_modules/sub/node_modules",
    "/home/user/git/main/node_modules",
    "/home/user/git/node_modules",
    "/home/user/node_modules",
    "/home/node_modules",
    "/node_modules"
]
```

Basically a predefined list of potential parent node_modules locations - perfect for use with `get-installed-path`.

The reason I need such a mechanism is that in some cases I have multiple projects using the same library which makes use of `get-installed-path`. Currently this system breaks due to npm optimising where the packages are installed (avoiding duplicates of course). Having this recursive behaviour would make this doable.

Fixes charlike/get-installed-path#11

### CODE REVIEW

**Comment by tunnckoCore** [comment]
Pretty cool, thanks! Best PR i've received.

It seems great. I'll merge later today and will apply some style fixes and prettier.

There are two modules by @jonschlinkert too that you might want to look: [npm-paths](https://github.com/jonschlinkert/npm-paths) and [global-paths](https://github.com/jonschlinkert/global-paths). In glance look output of both of them seems the same and like what `mainModule.paths` shows. But when look over the code, there are a bit differences (also some kind of support for electron).

Thanks, that feature definitely sounds good. Do you have some preferences for minimal node version, btw? Because i'm thinking of switching my modules to node >= v6. But there may be exceptions from the rule if you need lower version support.

**Comment by perry-mitchell** [comment]
> Pretty cool, thanks! Best PR i've received.

Thank you!

> There are two modules by @jonschlinkert too that you might want to look

Thanks for the links. It might even be a good idea to mention these in the docs for this library if you merge the `paths` support, imho. The reason I suggest using a user-providable array is **a.** the user can choose what to provide and **b.** the `process.mainModule.paths` approach should be called where the user intends (not in the `get-installed-paths` library, because of the entry script location).

That being said - I'm definitely open to searching for the best approach. Do you have any opinion on what should be done here? I'm using this library in my [force-latest-versions](https://github.com/perry-mitchell/force-latest-versions) project to locate versions of packages. Because of that I need the recursive upward searching capability.

> Do you have some preferences for minimal node version, btw? Because i'm thinking of switching my modules to node >= v6

I was actually considering the same for my project, but imho having 1 last version with this feature on maybe v4+ would be amazing. Perhaps you might consider reworking this project after that? 🙏

**Comment by tunnckoCore** [comment]
> The reason I suggest using a user-providable array is a. the user can choose what to provide and b. the process.mainModule.paths approach should be called where the user intends 

yea, absolutely, that's the correct way.

> The reason I suggest using a user-providable array is a. the user can choose what to provide and b. the process.mainModule.paths approach should be called where the user intends 

okey, no problems, i specifically don't need something special. I just considered to switch my account to org and i'm slowly updated **all** of my libs and move them at @tunnckoCore (i'll move this one too), so that's the time when i should unify all of them and decided to lock all to node >= 6, but of course accept exceptions, i need some too. Most are switching to v6 LTS, so in few months v4 won't be used so much.

Forcing people is terrible idea, yea. But forcing to use next LTS isn't really a force, imo :)

> Thanks for the links. It might even be a good idea to mention these in the docs 

of course, didn't thought for that.

**Comment by tunnckoCore** [comment]
Btw, forcing to latest versions using `latest` or `*` isn't good idea. it will break dependency keepers and updaters such as GK (which is terrible), but you should definitely see https://github.com/singapore/renovate and activate its Github App :) It has support for auto-update and auto-merge, have pretty granular control over configs and soon will support presets of configs :) It's one of the main reasons because i'm doing that jump to update all of my 300 packages in coming weeks.

**Comment by perry-mitchell** [comment]
> Btw, forcing to latest versions using latest or * isn't good idea

Absolutely agree here - I know the value behind good semver 😁. It's being used on an internal project which happens to require this method, but I'd not expect it to be used so much in the wild or with production projects.

EDIT: Come to think of it, this may be good as a disclaimer in my project.. heh.

**Comment by perry-mitchell** [comment]
@charlike please let me know if you'd like me to perform any changes!

**Comment by perry-mitchell** [comment]
@charlike please let me know if you'd like me to perform any changes!

**Comment by tunnckoCore** [comment]
> Come to think of it, this may be good as a disclaimer in my project.. heh.

Yea, that's what i was thinking to suggest you add some notes.

> please let me know if you'd like me to perform any changes!

No, it's okey. Mostly stylistic things, that will be fixed by my updates by adding eslint and prettier.
Don't mind :)

**Comment by perry-mitchell** [comment]
Hi @charlike - Have you made any decisions regarding this update?

**Comment by tunnckoCore** [comment]
Оh no no, sorry! Just forgot and was busy with other things. I'll get it finished in day or two max. :)

**Comment by tunnckoCore** [comment]
Hey, @perry-mitchell, it's now live as `2.1.0` version! Thanks for your contribution :tada: 

It is tested and works on Node 4, includes latest dependencies (the v1 of `global-modules` is important probably) and will be last v2 version. I'll publish new major soon too, that will require Node 6 and will test only on v6 and above on Travis. But it won't be something special and when you decide to migrate to v6 you can migrate to that major v3 release without any risks of any breaking changes.

**Comment by tunnckoCore** [comment]
Few more things to note about v3 release:
- it will use semantic-release
- use Github Releases instead of CHANGELOG
- will always be up-to-date, because everything on @tunnckoCore org is using RenovateApp for keeping (all) deps up to date, automatically and automagically with auto-publishing new release on any new release of the deps (which currently is only one - global-modules).


**Long story short:** in the moment that `global-modules` publish new release, in a few of hours after that  `get-installed-path` will receive that update and new patch version will be published (i even won't know)
Btw, this is true for **all** packages in that org ;p That automation is freaking awesome! :tada:
